drivers/can/can_stm32h7.c

```diff
@@ -0,0 +1,304 @@
+/*
+ * Copyright (c) 2021 Blue Clover
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/can.h>
+#include <drivers/pinctrl.h>
+#include <kernel.h>
+#include <soc.h>
+#include <stm32_ll_rcc.h>
+#include "can_stm32h7.h"
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(can_driver, CONFIG_CAN_LOG_LEVEL);
+
+#define DT_DRV_COMPAT st_stm32h7_fdcan
+
+int can_stm32h7_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	ARG_UNUSED(dev);
+	int rate_tmp;
+
+	rate_tmp = LL_RCC_GetFDCANClockFreq(LL_RCC_FDCAN_CLKSOURCE);
+
+	if (rate_tmp == LL_RCC_PERIPH_FREQUENCY_NO) {
+		LOG_ERR("Can't read core clock");
+		return -EIO;
+	}
+
+	*rate = rate_tmp;
+
+	LOG_DBG("rate=%d", *rate);
+
+	return 0;
+}
+
+int can_stm32h7_get_max_filters(const struct device *dev, enum can_ide id_type)
+{
+	if (id_type == CAN_STANDARD_IDENTIFIER) {
+		return NUM_STD_FILTER_DATA;
+	} else {
+		return NUM_EXT_FILTER_DATA;
+	}
+}
+
+void can_stm32h7_clock_enable(void)
+{
+	LL_RCC_SetFDCANClockSource(LL_RCC_FDCAN_CLKSOURCE_PLL1Q);
+	__HAL_RCC_FDCAN_CLK_ENABLE();
```

Comment:
Please use existing clock_control API, rather than HAL direct access.
(I know this is the same as used in can_stm32fd, but this one also has to be fixed)

Reply:
@erwango Now that the rest of the PR is pretty much wrapped, I'm looking at this. I'm not sure exactly what all needs to be done to switch to clock_control for this. It doesn't look like this functionality is in clock_control yet, unless I'm missing something.

I'm not familiar enough with clock_control to know where to start with this.

Reply:
@erwango Thanks for the pointers—I've switched to using `clock_control_on()` to turn on the FDCAN peripheral clock. As we discussed, the clock selection for FDCAN is still using LL calls, but at least there is no HAL stuff anymore.

I think this resolves this for now, until there is support for controlling FDCAN clock selection in clock_control.